finding(plugin-dashboard): ObjectDataTable's `finalData` fallback re-derives every column on every render while no rows have arrived

Observed while fixing #4618 (PR #4623), which traced the same expression shape one file over. Filed unassigned; out of that card's scope — #4618 is the STATIC `table` path, this is the `provider: 'object'` sibling. Duplicate-searched (keyword + `ObjectDataTable` / `finalData` path): nothing open. The closest relatives are both closed — #4567 (a columns-array identity in a fetch dependency chain) and #4618 itself.

## What

`packages/plugin-dashboard/src/ObjectDataTable.tsx:245`

```ts
const finalData = Array.isArray(rawData) ? rawData : [];
```

The non-array fallback is a fresh array literal per render, and `finalData` is a dependency of the `derivedColumns` memo (`:332`). So while `rawData` is anything other than an array — the whole pre-fetch window, and any render where the provider hands back a non-array — the memo's key changes on every render and `derivedColumns` recomputes: `buildFieldMeta` per column, a fresh `cell` closure per column, the `isSystemField` pass, the `fieldLabel` lookups. All of it discarded, because the memo returns `[]` at `:320` (`if (finalData.length === 0) return []`) before any of it is used.

ESLint already reports it and has all along:

```
warning  The 'finalData' conditional could make the dependencies of useMemo Hook (at line 332)
         change on every render. To fix this, wrap the initialization of 'finalData' in its own
         useMemo() Hook          react-hooks/exhaustive-deps
```

## Why this is observation-class, not a defect a user hits

Named explicitly so triage does not have to re-derive it: this **cannot** produce #4618's render loop. That loop needed the churn to feed a `setState`, and it needed the state write to re-render the component that regenerates the literal. Here the churn feeds a memo only. It reaches `data-table`'s `columns` prop, but `data-table` re-rendering does not re-render `ObjectDataTable`, so nothing sustains it — and since PR #4623 the columns sync compares by value, so the churn no longer costs even the one extra child render it used to.

What is left is wasted work in the loading window, and one live ESLint warning that documents a real identity bug. Nothing renders wrong.

## Fix shape

The same one-line move PR #4623 made in `data-table.tsx`: a module-scope frozen empty instead of a per-render literal, so "no rows yet" is a stable value. `ObjectDataTable.tsx:245` is the only site; `schema.data` and `boundData` are already stable when present.

Worth pairing with a sweep for the same expression across the package — `Array.isArray(x) ? x : []` inside a component body is the shape, and it is only ever safe when nothing memoizes or synchronizes on the result.

---
_Generated by [Claude Code](https://claude.ai/code/session_017Qqyix2QcnpUC9XeYVDzx3)_

<!-- os-dev-report -->

```json
{
  "issue": 4629,
  "status": "done",
  "branch": "claude/issue-4629-objectdatatable-stable-empty",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5566",
  "premise_still_valid": true,
  "head_sha": "c8fd1be97",
  "summary": "Both ObjectDataTable.tsx:310 and ObjectPivotTable.tsx:154 now fall back to a module-scope Object.freeze([]) instead of a per-render literal, the same shape data-table.tsx uses for EMPTY_ROWS (#4618/PR #4623); no useMemo wrapper and no memo dependency-list change. Two sub-claims of the card were corrected by measurement without disturbing its premise: the churn is NOT the whole pre-fetch window (fetchedData is a useState-seeded array with stable identity, so a truthy non-array boundData/schema.data is required), and the render-loop impossibility holds by a stronger argument than the card gave (in ObjectDataTable the fallback fires iff finalData.length===0, which is exactly the condition under which the component early-returns and never renders SchemaRenderer, so the churn never reaches data-table at all). The sweep found one site the card and dispatch both missed (PivotTable.tsx, ESLint-confirmed) and falsified one the dispatch listed as in-surface (ObjectDataTable.tsx:162 is not a component body); both are itemised below.",
  "files_changed": [
    "packages/plugin-dashboard/src/ObjectDataTable.tsx",
    "packages/plugin-dashboard/src/ObjectPivotTable.tsx",
    "packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx",
    "packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx",
    ".changeset/dashboard-stable-empty-rows-4629.md"
  ],
  "gates": [
    { "gate": "pnpm exec turbo run lint --filter=@object-ui/plugin-dashboard", "verdict": "PASS", "printed": "Tasks: 2 successful, 2 total; eslint: 0 errors, 360 warnings" },
    { "gate": "pnpm exec turbo run type-check --filter=@object-ui/plugin-dashboard", "verdict": "PASS", "printed": "Tasks: 13 successful, 13 total (incl. tsc --noEmit && tsc -p tsconfig.test.json)" },
    { "gate": "pnpm exec vitest run packages/plugin-dashboard/", "verdict": "PASS", "printed": "Test Files 72 passed (72) / Tests 653 passed (653)" },
    { "gate": "node scripts/check-control-bytes.mjs", "verdict": "PASS", "printed": "check-control-bytes: OK (scanned 4620 tracked text file(s); skipped 85 binary)." },
    { "gate": "node scripts/check-changeset-presence.mjs", "verdict": "PASS", "printed": "4 source file(s) of 1 released package(s) changed, and this change declares 1 changeset(s)" },
    { "gate": "node scripts/check-changeset-no-major.mjs", "verdict": "PASS", "printed": "No changeset declares a `major` bump." },
    { "gate": "node scripts/check-changeset-fixed.mjs", "verdict": "PASS", "printed": "All workspace packages are in the changeset fixed group." },
    { "gate": "node scripts/check-lint-coverage.mjs", "verdict": "PASS", "printed": "lint coverage: 46/46 packages linted, 0 with outstanding errors (0 total)." },
    { "gate": "node scripts/check-type-check-coverage.mjs", "verdict": "PASS", "printed": "test type-check coverage: 41/41 packages compile their tests, 0 declared debt" },
    { "gate": "pnpm check:self-import", "verdict": "PASS", "printed": "No package names itself inside its own src/." },
    { "gate": "pnpm check:phantom-deps", "verdict": "PASS", "printed": "Every in-scope import is declared by the package that publishes it." }
  ],
  "gates_not_run": [
    { "gate": "pnpm lint (repo-wide, 46 packages)", "reason": "DECLARED NARROWING to this package's own `eslint .` task. Evidence: population is eslint's own config resolution (the package script is literally `eslint .`); 102 files linted read from --format json output length; eslint.config.js enables no type-aware linting (no parserOptions.project / projectService / *TypeChecked) and none of the 7 rules in eslint-rules/ reads any file but the one under lint, so this diff cannot move any untouched file's verdict. check-lint-coverage.mjs independently reports 46/46 packages, 0 outstanding errors." },
    { "gate": "pnpm check (CLI self-check), doc-component-types, doc-snippet-types, docs-links, skills-paths, published-dist, e2e", "reason": "CI-owned; diff touches no CLI, docs, skills/, dist tooling or e2e surface." }
  ],
  "tests": "New: ObjectDataTable.stableEmptyRows.test.tsx (recompute count via a partial vi.mock of ../recordFields counting buildFieldMeta calls; load-bearing mount pre-assertion toBe(2) so it cannot pass vacuously; also asserts the table is NOT rendered in that window) and ObjectPivotTable.stableEmptyRows.test.tsx (direct identity: new Set(capturedData).size === 1 across mount + 3 host re-renders, plus Object.isFrozen). Post-fix: `pnpm exec vitest run <both files>` -> `Test Files 2 passed (2) / Tests 4 passed (4)`. Whole package post-fix: `Test Files 72 passed (72) / Tests 653 passed (653)`. REVERSE VERIFICATION: directions written before the run (ObjectDataTable RED, ObjectPivotTable RED, both negative controls GREEN on both sides) and all matched -- measured `AssertionError: expected 8 to be 2` and `AssertionError: expected 4 to be 1`, negative controls `2 passed`. No rebuild leg applies: these suites import ../ObjectDataTable and ../ObjectPivotTable from src directly and vitest resolves @object-ui/* to src via the root config alias table, so no dist is in the resolution path. Mutation confirmed ON DISK by anchored grep -c on the reverted text ('rawData : EMPTY_ROWS' 0/0, 'rawData : []' 1/1, '^const EMPTY_ROWS' 0/0) -- NOT by an editor exit code, and NOT by `git diff --stat`, which was empty because `git checkout <ref> -- <paths>` updates the index too. The ablation script carried `trap '<restore>' EXIT INT TERM`; the tree was restored byte-identically (`git diff --exit-code HEAD` clean, `git status --short` empty). ESLint before/after measured in a second trap-guarded run: origin/main printed `ObjectDataTable.tsx:310:9 warning The 'finalData' conditional could make the dependencies of useMemo Hook (at line 417) change on every render ... react-hooks/exhaustive-deps`; the branch prints zero such warnings.",
  "sweep_disposition": [
    { "site": "packages/plugin-dashboard/src/ObjectDataTable.tsx:310 finalData", "disposition": "FIXED", "measurement": "keys the derivedColumns memo (dep list at :417); ESLint-reported on origin/main; early return at :404 discards the result" },
    { "site": "packages/plugin-dashboard/src/ObjectPivotTable.tsx:154 finalData", "disposition": "FIXED", "measurement": "handed to PivotTable as finalSchema.data, keys its cross-tab memo (deps [data,rowField,columnField,valueField,aggregation] at PivotTable.tsx:241); no ESLint warning here because it leaves the file via props" },
    { "site": "packages/plugin-dashboard/src/ObjectDataTable.tsx:162 cols", "disposition": "SKIPPED - NOT the defect shape (dispatch listed it as in-surface; falsified)", "measurement": "inside `export function computeLookupExpand`, a module-scope pure function, not a component body. `cols` is used only within the call (.length, .map) and never returned; the function's own result is a fresh Array.from(out) per call regardless; its single call site (:263) is a fetch-effect body, not a dependency list. Nothing memoizes or synchronizes on it, so by the card's own criterion it is safe." },
    { "site": "packages/plugin-dashboard/src/PivotTable.tsx:176 data, plus the `data: rawData = []` destructuring default at ~:146", "disposition": "SKIPPED - out of declared file surface; FILED as #5562", "measurement": "NEW site, in neither the card nor the dispatch. ESLint on origin/main: \"The 'data' conditional could make the dependencies of useMemo Hook (at line 241) change on every render\". This PR closes the object-bound path through ObjectPivotTable (the stable empty passes through the Array.isArray arm unchanged); the direct-use path remains." },
    { "site": "packages/plugin-dashboard/src/DashboardFilterBar.tsx:300", "disposition": "SKIPPED", "measurement": "inside a .then() callback; result feeds setDynamicOptions(pairOptionRows(rows, rawRows, from)) - a setState payload, not a per-render literal" },
    { "site": "packages/plugin-dashboard/src/DashboardWithConfig.tsx:111-112", "disposition": "SKIPPED", "measurement": "inside `selectedWidgetConfig = React.useMemo(...)`, deps [selectedWidgetId, configVersion]" },
    { "site": "packages/plugin-dashboard/src/DatasetWidget.tsx:400-401", "disposition": "SKIPPED", "measurement": "each already wrapped in its own useMemo, deps [widget]" },
    { "site": "packages/plugin-dashboard/src/DatasetWidget.tsx:529", "disposition": "SKIPPED", "measurement": "inside a .then(); a setState payload" },
    { "site": "packages/plugin-dashboard/src/ObjectMetricWidget.tsx:253", "disposition": "SKIPPED", "measurement": "inside an async useCallback body (computeOne); local to the call" }
  ],
  "deviations": [
    "The dispatch's gate list was inapplicable: it was produced by objectstack's scripts/pm/dispatch-gates.mjs, which resolves its git cwd to its OWN repo root, so it matched objectui paths against objectstack's globs. I measured that 9 of the 10 named gates do not exist in objectui (only scripts/check-changeset-no-major.mjs does) and derived objectui's real families from its package.json and .github/workflows instead. The PM sent the same correction mid-task; the two derivations agree.",
    "The dispatch classified ObjectDataTable.tsx:162 as in-surface 'same defect shape'. Measured otherwise (not a component body, nothing memoizes on it) and left unchanged rather than applying a cosmetic hoist. Detailed above and in the PR body.",
    "PivotTable.tsx carries the same defect and was left untouched to respect the declared file surface; filed as #5562 rather than fixed in place. The bounded-in-place-fix exemption was not taken because the card clause naming the file surface is explicit and PivotTable is a third component with its own entry path that would need its own test surface.",
    "ObjectDataTable.stableEmptyRows.test.tsx renders inside a real I18nProvider. This is load-bearing, not decoration: without an i18next instance the memo's fieldLabel/fieldOptionLabel deps churn independently (measured 8 buildFieldMeta calls, unchanged by this fix), which would have made the test green pre-fix in the wrong direction. Root cause measured and filed as #5564.",
    "objectui's git checkout is shallow, so `dispatch-gates.mjs` refused to derive a change set from cwd. `git fetch --deepen=200 origin main` did not clear the shallow flag; the change set was instead derived explicitly from `git merge-base origin/main HEAD` (0935a43be), never from the two-dot origin/main..HEAD range."
  ],
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #5562: PivotTable.tsx carries two per-render empty-array literals (the `data: rawData = []` destructuring default and the `Array.isArray(rawData) ? rawData : []` at :176) that both key its cross-tabulation memo at :241 - ESLint-confirmed on origin/main, same class as #4629, outside this card's file surface",
    "filed as #5564: outside an i18next instance, useSafeFieldLabel returns a FRESH object every render (measured 4 distinct identities in 4 renders; 1 when an instance exists), because useObjectLabel memoizes on react-i18next's [t, i18n] and react-i18next hands back a fresh t with no instance to bind to - so the memo protection lapses in exactly the no-provider case the wrapper's docstring says it exists to serve"
  ]
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

os-dev-report

**This supersedes comment 5370027634.** That one was posted with the marker as an HTML comment and the GitHub body sanitizer ate it after storage (verified by read-back: the stored body begins with a blank line, no marker), which makes it invisible to a marker scan. It also swallowed four `<…>` fragments inside the `tests` string. This copy uses the literal-text marker and spells those fragments without angle brackets; the content is otherwise identical.

```json
{
  "issue": 4629,
  "status": "done",
  "branch": "claude/issue-4629-objectdatatable-stable-empty",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5566",
  "premise_still_valid": true,
  "head_sha": "c8fd1be97",
  "summary": "Both ObjectDataTable.tsx:310 and ObjectPivotTable.tsx:154 now fall back to a module-scope Object.freeze([]) instead of a per-render literal, the same shape data-table.tsx uses for EMPTY_ROWS (#4618/PR #4623); no useMemo wrapper and no memo dependency-list change. Two sub-claims of the card were corrected by measurement without disturbing its premise: the churn is NOT the whole pre-fetch window (fetchedData is a useState-seeded array with stable identity, so a truthy non-array boundData/schema.data is required), and the render-loop impossibility holds by a stronger argument than the card gave (in ObjectDataTable the fallback fires iff finalData.length===0, which is exactly the condition under which the component early-returns and never renders SchemaRenderer, so the churn never reaches data-table at all). The sweep found one site the card and dispatch both missed (PivotTable.tsx, ESLint-confirmed) and falsified one the dispatch listed as in-surface (ObjectDataTable.tsx:162 is not a component body); both are itemised below.",
  "files_changed": [
    "packages/plugin-dashboard/src/ObjectDataTable.tsx",
    "packages/plugin-dashboard/src/ObjectPivotTable.tsx",
    "packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx",
    "packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx",
    ".changeset/dashboard-stable-empty-rows-4629.md"
  ],
  "gates": [
    { "gate": "pnpm exec turbo run lint --filter=@object-ui/plugin-dashboard", "verdict": "PASS", "printed": "Tasks: 2 successful, 2 total; eslint: 0 errors, 360 warnings" },
    { "gate": "pnpm exec turbo run type-check --filter=@object-ui/plugin-dashboard", "verdict": "PASS", "printed": "Tasks: 13 successful, 13 total (incl. tsc --noEmit && tsc -p tsconfig.test.json)" },
    { "gate": "pnpm exec vitest run packages/plugin-dashboard/", "verdict": "PASS", "printed": "Test Files 72 passed (72) / Tests 653 passed (653)" },
    { "gate": "node scripts/check-control-bytes.mjs", "verdict": "PASS", "printed": "check-control-bytes: OK (scanned 4620 tracked text file(s); skipped 85 binary)." },
    { "gate": "node scripts/check-changeset-presence.mjs", "verdict": "PASS", "printed": "4 source file(s) of 1 released package(s) changed, and this change declares 1 changeset(s)" },
    { "gate": "node scripts/check-changeset-no-major.mjs", "verdict": "PASS", "printed": "No changeset declares a `major` bump." },
    { "gate": "node scripts/check-changeset-fixed.mjs", "verdict": "PASS", "printed": "All workspace packages are in the changeset fixed group." },
    { "gate": "node scripts/check-lint-coverage.mjs", "verdict": "PASS", "printed": "lint coverage: 46/46 packages linted, 0 with outstanding errors (0 total)." },
    { "gate": "node scripts/check-type-check-coverage.mjs", "verdict": "PASS", "printed": "test type-check coverage: 41/41 packages compile their tests, 0 declared debt" },
    { "gate": "pnpm check:self-import", "verdict": "PASS", "printed": "No package names itself inside its own src/." },
    { "gate": "pnpm check:phantom-deps", "verdict": "PASS", "printed": "Every in-scope import is declared by the package that publishes it." }
  ],
  "gates_not_run": [
    { "gate": "pnpm lint (repo-wide, 46 packages)", "reason": "DECLARED NARROWING to this package's own `eslint .` task. Evidence: population is eslint's own config resolution (the package script is literally `eslint .`); 102 files linted read from --format json output length; eslint.config.js enables no type-aware linting (no parserOptions.project / projectService / *TypeChecked) and none of the 7 rules in eslint-rules/ reads any file but the one under lint, so this diff cannot move any untouched file's verdict. check-lint-coverage.mjs independently reports 46/46 packages, 0 outstanding errors." },
    { "gate": "pnpm check (CLI self-check), doc-component-types, doc-snippet-types, docs-links, skills-paths, published-dist, e2e", "reason": "CI-owned; diff touches no CLI, docs, skills/, dist tooling or e2e surface." }
  ],
  "tests": "New: ObjectDataTable.stableEmptyRows.test.tsx (recompute count via a partial vi.mock of ../recordFields counting buildFieldMeta calls; load-bearing mount pre-assertion toBe(2) so it cannot pass vacuously; also asserts the table is NOT rendered in that window) and ObjectPivotTable.stableEmptyRows.test.tsx (direct identity: new Set(capturedData).size === 1 across mount + 3 host re-renders, plus Object.isFrozen). Post-fix: `pnpm exec vitest run` on the two new files -> `Test Files 2 passed (2) / Tests 4 passed (4)`. Whole package post-fix: `Test Files 72 passed (72) / Tests 653 passed (653)`. REVERSE VERIFICATION: directions written before the run (ObjectDataTable RED, ObjectPivotTable RED, both negative controls GREEN on both sides) and all matched -- measured `AssertionError: expected 8 to be 2` and `AssertionError: expected 4 to be 1`, negative controls `2 passed`. No rebuild leg applies: these suites import ../ObjectDataTable and ../ObjectPivotTable from src directly and vitest resolves @object-ui/* to src via the root config alias table, so no dist is in the resolution path. Mutation confirmed ON DISK by anchored grep -c on the reverted text ('rawData : EMPTY_ROWS' 0/0, 'rawData : []' 1/1, '^const EMPTY_ROWS' 0/0) -- NOT by an editor exit code, and NOT by `git diff --stat`, which was empty because `git checkout REF -- PATHS` updates the index too. The ablation script carried a `trap` on EXIT INT TERM that restores the files; the tree was restored byte-identically (`git diff --exit-code HEAD` clean, `git status --short` empty). ESLint before/after measured in a second trap-guarded run: origin/main printed `ObjectDataTable.tsx:310:9 warning The 'finalData' conditional could make the dependencies of useMemo Hook (at line 417) change on every render ... react-hooks/exhaustive-deps`; the branch prints zero such warnings.",
  "sweep_disposition": [
    { "site": "packages/plugin-dashboard/src/ObjectDataTable.tsx:310 finalData", "disposition": "FIXED", "measurement": "keys the derivedColumns memo (dep list at :417); ESLint-reported on origin/main; early return at :404 discards the result" },
    { "site": "packages/plugin-dashboard/src/ObjectPivotTable.tsx:154 finalData", "disposition": "FIXED", "measurement": "handed to PivotTable as finalSchema.data, keys its cross-tab memo (deps [data,rowField,columnField,valueField,aggregation] at PivotTable.tsx:241); no ESLint warning here because it leaves the file via props" },
    { "site": "packages/plugin-dashboard/src/ObjectDataTable.tsx:162 cols", "disposition": "SKIPPED - NOT the defect shape (dispatch listed it as in-surface; falsified)", "measurement": "inside `export function computeLookupExpand`, a module-scope pure function, not a component body. `cols` is used only within the call (.length, .map) and never returned; the function's own result is a fresh Array.from(out) per call regardless; its single call site (:263) is a fetch-effect body, not a dependency list. Nothing memoizes or synchronizes on it, so by the card's own criterion it is safe." },
    { "site": "packages/plugin-dashboard/src/PivotTable.tsx:176 data, plus the `data: rawData = []` destructuring default at ~:146", "disposition": "SKIPPED - out of declared file surface; FILED as #5562", "measurement": "NEW site, in neither the card nor the dispatch. ESLint on origin/main: \"The 'data' conditional could make the dependencies of useMemo Hook (at line 241) change on every render\". This PR closes the object-bound path through ObjectPivotTable (the stable empty passes through the Array.isArray arm unchanged); the direct-use path remains." },
    { "site": "packages/plugin-dashboard/src/DashboardFilterBar.tsx:300", "disposition": "SKIPPED", "measurement": "inside a .then() callback; result feeds setDynamicOptions(pairOptionRows(rows, rawRows, from)) - a setState payload, not a per-render literal" },
    { "site": "packages/plugin-dashboard/src/DashboardWithConfig.tsx:111-112", "disposition": "SKIPPED", "measurement": "inside `selectedWidgetConfig = React.useMemo(...)`, deps [selectedWidgetId, configVersion]" },
    { "site": "packages/plugin-dashboard/src/DatasetWidget.tsx:400-401", "disposition": "SKIPPED", "measurement": "each already wrapped in its own useMemo, deps [widget]" },
    { "site": "packages/plugin-dashboard/src/DatasetWidget.tsx:529", "disposition": "SKIPPED", "measurement": "inside a .then(); a setState payload" },
    { "site": "packages/plugin-dashboard/src/ObjectMetricWidget.tsx:253", "disposition": "SKIPPED", "measurement": "inside an async useCallback body (computeOne); local to the call" }
  ],
  "deviations": [
    "The dispatch's gate list was inapplicable: it was produced by objectstack's scripts/pm/dispatch-gates.mjs, which resolves its git cwd to its OWN repo root, so it matched objectui paths against objectstack's globs. I measured that 9 of the 10 named gates do not exist in objectui (only scripts/check-changeset-no-major.mjs does) and derived objectui's real families from its package.json and .github/workflows instead. The PM sent the same correction mid-task; the two derivations agree.",
    "The dispatch classified ObjectDataTable.tsx:162 as in-surface 'same defect shape'. Measured otherwise (not a component body, nothing memoizes on it) and left unchanged rather than applying a cosmetic hoist. Detailed above and in the PR body.",
    "PivotTable.tsx carries the same defect and was left untouched to respect the declared file surface; filed as #5562 rather than fixed in place. The bounded-in-place-fix exemption was not taken because the card clause naming the file surface is explicit and PivotTable is a third component with its own entry path that would need its own test surface.",
    "ObjectDataTable.stableEmptyRows.test.tsx renders inside a real I18nProvider. This is load-bearing, not decoration: without an i18next instance the memo's fieldLabel/fieldOptionLabel deps churn independently (measured 8 buildFieldMeta calls, unchanged by this fix), which would have made the test green pre-fix in the wrong direction. Root cause measured and filed as #5564.",
    "objectui's git checkout is shallow, so `dispatch-gates.mjs` refused to derive a change set from cwd. `git fetch --deepen=200 origin main` did not clear the shallow flag; the change set was instead derived explicitly from `git merge-base origin/main HEAD` (0935a43be), never from the two-dot origin/main..HEAD range.",
    "The first delivery of this report (comment 5370027634) lost its `os-dev-report` HTML-comment marker to the GitHub body sanitizer; this comment re-delivers it with a literal-text marker."
  ],
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #5562: PivotTable.tsx carries two per-render empty-array literals (the `data: rawData = []` destructuring default and the `Array.isArray(rawData) ? rawData : []` at :176) that both key its cross-tabulation memo at :241 - ESLint-confirmed on origin/main, same class as #4629, outside this card's file surface",
    "filed as #5564: outside an i18next instance, useSafeFieldLabel returns a FRESH object every render (measured 4 distinct identities in 4 renders; 1 when an instance exists), because useObjectLabel memoizes on react-i18next's [t, i18n] and react-i18next hands back a fresh t with no instance to bind to - so the memo protection lapses in exactly the no-provider case the wrapper's docstring says it exists to serve"
  ]
}
```


---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/dashboard-stable-empty-rows-4629.md
+++ b/.changeset/dashboard-stable-empty-rows-4629.md
@@ -1,0 +1,22 @@
+---
+'@object-ui/plugin-dashboard': patch
+---
+
+`ObjectDataTable` and `ObjectPivotTable` now use a module-scope frozen empty for
+"no rows yet" instead of a fresh array literal per render (objectui#4629).
+
+Both spelled the resolved row list as `Array.isArray(rawData) ? rawData : []`, so
+whenever `rawData` was a truthy non-array — a provider-config `data`, or a `bind`
+path that resolves to an object — the fallback produced a NEW array identity on
+every render. In `ObjectDataTable` that value keys the `derivedColumns` memo, so
+every column was re-derived (`buildFieldMeta`, a fresh `cell` closure, the
+`isSystemField` pass, the `fieldLabel` lookups) and then discarded by the
+`finalData.length === 0` early return. In `ObjectPivotTable` the value is handed
+straight to `PivotTable`, where it keys the cross-tabulation memo, so the pivot
+rebuilt its row/column sets, bucket map and totals on every render over no rows
+at all.
+
+Nothing rendered wrong before or after; this is wasted work in the empty window,
+plus the live `react-hooks/exhaustive-deps` warning the conditional raised. It is
+the same module-scope frozen empty `data-table.tsx` adopted for its own
+`EMPTY_ROWS` (objectui#4618), applied to the `provider: 'object'` siblings.

--- a/packages/plugin-dashboard/src/ObjectDataTable.tsx
+++ b/packages/plugin-dashboard/src/ObjectDataTable.tsx
@@ -47,6 +47,29 @@ interface NormalizedColumn {
 }
 
 /**
+ * Shared empty fallback for the resolved row list (objectui#4629).
+ *
+ * `Array.isArray(rawData) ? rawData : []` evaluates a FRESH array on every
+ * render, and that value is a dependency of the `derivedColumns` memo below.
+ * So for as long as `rawData` is a non-array — a provider-config `data`, or a
+ * `bind` path that resolves to an object — the memo's key changes on every
+ * render and every column is re-derived (`buildFieldMeta`, a fresh `cell`
+ * closure per column, the `isSystemField` pass, the `fieldLabel` lookups) only
+ * to be discarded: `finalData.length === 0` is exactly the case in which the
+ * component returns its empty state without ever rendering the table.
+ * Hoisting the empty to module scope makes "no rows yet" a stable value, so
+ * the memo sees what is actually true — nothing changed.
+ *
+ * The same move `data-table.tsx` made for its own `EMPTY_ROWS` (objectui#4618,
+ * PR #4623). This file is the `provider: 'object'` sibling of that one, so it
+ * takes the same shape rather than a second remedy for one defect class.
+ *
+ * Frozen so a consumer that mutates the array it was handed cannot corrupt the
+ * shared instance for every other table on the page.
+ */
+const EMPTY_ROWS = Object.freeze([]) as unknown as any[];
+
+/**
  * Normalize columns to support both string[] shorthand and object[] formats.
  *
  * - `string[]` entries are converted to `{ header, accessorKey }` objects,
@@ -307,7 +330,7 @@ export const ObjectDataTable: React.FC<ObjectDataTableProps> = ({ schema, dataSo
 
   // Resolve data: bound data > static schema data > fetched data
   const rawData = boundData || schema.data || fetchedData;
-  const finalData = Array.isArray(rawData) ? rawData : [];
+  const finalData = Array.isArray(rawData) ? rawData : EMPTY_ROWS;
 
   // Auto-derive columns from data keys when none are provided. When `objectName`
   // is set, prefer translated field labels via the convention-based hook so that

--- a/packages/plugin-dashboard/src/ObjectPivotTable.tsx
+++ b/packages/plugin-dashboard/src/ObjectPivotTable.tsx
@@ -16,6 +16,28 @@ import { DrillDownDrawer } from './DrillDownDrawer';
 import { resolveFilterPlaceholders } from './utils';
 import type { PivotTableSchema } from '@object-ui/types';
 
+/**
+ * Shared empty fallback for the resolved row list (objectui#4629).
+ *
+ * `Array.isArray(rawData) ? rawData : []` evaluates a FRESH array on every
+ * render, and this component hands the result straight to `PivotTable` as
+ * `finalSchema.data`, where it is the identity key of the cross-tabulation
+ * memo (`PivotTable.tsx`, `[data, rowField, columnField, valueField,
+ * aggregation]`). So while `rawData` is a non-array — a provider-config
+ * `data`, or a `bind` path that resolves to an object — that memo rebuilds its
+ * row/column sets, bucket map and totals on every render over nothing at all.
+ * A module-scope empty makes "no rows yet" a stable value, and the
+ * `Array.isArray` arm downstream passes it through unchanged, so the memo
+ * holds.
+ *
+ * The same move `data-table.tsx` made for its own `EMPTY_ROWS` (objectui#4618,
+ * PR #4623) and `ObjectDataTable.tsx` makes for the identical expression.
+ *
+ * Frozen so a consumer that mutates the array it was handed cannot corrupt the
+ * shared instance for every other pivot on the page.
+ */
+const EMPTY_ROWS = Object.freeze([]) as unknown as any[];
+
 export interface ObjectPivotTableProps {
   schema: PivotTableSchema & {
     objectName?: string;
@@ -151,7 +173,7 @@ export const ObjectPivotTable: React.FC<ObjectPivotTableProps> = ({ schema, data
 
   // Resolve data: bound data > static schema data > fetched data
   const rawData = boundData || schema.data || fetchedData;
-  const finalData = Array.isArray(rawData) ? rawData : [];
+  const finalData = Array.isArray(rawData) ? rawData : EMPTY_ROWS;
 
   // Loading skeleton
   if (loading && finalData.length === 0) {

--- a/packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx
+++ b/packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx
@@ -103,8 +103,9 @@ const I18N_CONFIG = { defaultLanguage: 'en', detectBrowserLanguage: false } as c
  * `finalData` is the only churning dependency and this test measures exactly
  * it. OUTSIDE any provider react-i18next has no instance to bind to and hands
  * back a fresh `t` every render, so the memo re-keys regardless of this fix —
- * measured on this branch (8 calls, unchanged by the fix) and filed separately;
- * it is a defect in a different package and is not what #4629 is about.
+ * measured on this branch (8 calls, unchanged by the fix) and filed as
+ * objectui#5564; it is a defect in a different package and is not what #4629
+ * is about.
  */
 function renderUnderHostChurn(schema: any) {
   let bump: (() => void) | null = null;

--- a/packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx
+++ b/packages/plugin-dashboard/src/__tests__/ObjectDataTable.stableEmptyRows.test.tsx
@@ -1,0 +1,167 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * objectui#4629 — "no rows yet" must be a STABLE value in `ObjectDataTable`.
+ *
+ * `const finalData = Array.isArray(rawData) ? rawData : []` evaluated a FRESH
+ * array literal on every render, and `finalData` is a dependency of the
+ * `derivedColumns` memo. So while `rawData` is a non-array the memo's key
+ * changed on every render and every column was re-derived — `buildFieldMeta`,
+ * a fresh `cell` closure, the `isSystemField` pass, the `fieldLabel` lookups —
+ * and then discarded, because `finalData.length === 0` is exactly the case in
+ * which the component returns its empty state without rendering the table.
+ *
+ * Nothing renders WRONG either before or after, so a "the table renders
+ * correctly" assertion is green against the broken code and proves nothing.
+ * The observable that separates the two is the RECOMPUTE COUNT: one
+ * `buildFieldMeta` call per declared column per memo evaluation. An
+ * `eslint-disable` suppression of the `react-hooks/exhaustive-deps` warning
+ * would silence the warning and leave this count growing, so this test
+ * discriminates the real fix from a suppression.
+ *
+ * WHICH rawData is a non-array, measured rather than assumed: `rawData` is
+ * `boundData || schema.data || fetchedData`, and `fetchedData` is
+ * `useState<any[]>([])` — an array with a STABLE identity from the first
+ * render. So the plain pre-fetch window was never the churning case; the
+ * churn needs a truthy non-array, i.e. a provider-config `data` (below) or a
+ * `bind` path that resolves to an object.
+ */
+import { describe, it, expect, afterEach, vi } from 'vitest';
+import { render, cleanup, act } from '@testing-library/react';
+import { I18nProvider } from '@object-ui/i18n';
+import React from 'react';
+
+const counters = vi.hoisted(() => ({ buildFieldMeta: 0 }));
+
+// Count memo evaluations at the per-column work the memo does. `importActual`
+// keeps the real enrichment running, so the component still behaves normally.
+vi.mock('../recordFields', async () => {
+  const actual: any = await vi.importActual('../recordFields');
+  return {
+    ...actual,
+    buildFieldMeta: (...args: any[]) => {
+      counters.buildFieldMeta += 1;
+      return actual.buildFieldMeta(...args);
+    },
+  };
+});
+
+// The underlying `data-table` renderer is stubbed so the test does not depend
+// on the component registry — the same shape the sibling suites in this
+// directory use.
+vi.mock('@object-ui/react', async () => {
+  const actual: any = await vi.importActual('@object-ui/react');
+  return {
+    ...actual,
+    SchemaRenderer: ({ schema }: any) => (
+      <div data-testid="table">
+        {(schema.data ?? []).map((row: any, i: number) => (
+          <span key={i}>{String(row.name)}</span>
+        ))}
+      </div>
+    ),
+  };
+});
+
+import { ObjectDataTable } from '../ObjectDataTable';
+
+afterEach(() => {
+  cleanup();
+  counters.buildFieldMeta = 0;
+});
+
+/**
+ * A schema whose `data` is a PROVIDER CONFIG object rather than rows — the
+ * same case `data-table.tsx`'s own `EMPTY_ROWS` docstring names. `columns` is
+ * declared so the memo takes its full enrichment branch: this is the window
+ * where the work is done and thrown away.
+ */
+const PROVIDER_CONFIG_SCHEMA = {
+  type: 'object-table',
+  data: { provider: 'object', object: 'account' },
+  columns: ['name', 'amount'],
+} as any;
+
+/**
+ * Declared at module scope, not inline: `I18nProvider` memoizes its i18next
+ * bootstrap on the `config` IDENTITY, so an inline literal would rebuild the
+ * instance every render — the very class of bug under test, one level up.
+ */
+const I18N_CONFIG = { defaultLanguage: 'en', detectBrowserLanguage: false } as const;
+
+/**
+ * The widget is rendered inside a real `I18nProvider`, which is load-bearing
+ * rather than decoration. `derivedColumns` also depends on `fieldLabel` /
+ * `fieldOptionLabel` from `useSafeFieldLabel`, which memoize on react-i18next's
+ * `[t, i18n]`. With a real instance those are stable between renders, so
+ * `finalData` is the only churning dependency and this test measures exactly
+ * it. OUTSIDE any provider react-i18next has no instance to bind to and hands
+ * back a fresh `t` every render, so the memo re-keys regardless of this fix —
+ * measured on this branch (8 calls, unchanged by the fix) and filed separately;
+ * it is a defect in a different package and is not what #4629 is about.
+ */
+function renderUnderHostChurn(schema: any) {
+  let bump: (() => void) | null = null;
+  const Host = () => {
+    const [, setTick] = React.useState(0);
+    bump = () => setTick((n) => n + 1);
+    return <ObjectDataTable schema={schema} />;
+  };
+  const utils = render(<I18nProvider config={I18N_CONFIG}><Host /></I18nProvider>);
+  return {
+    ...utils,
+    churn: (times: number) => {
+      for (let i = 0; i < times; i += 1) act(() => { bump!(); });
+    },
+  };
+}
+
+describe('ObjectDataTable keeps "no rows yet" stable (#4629)', () => {
+  it('derives each column once, however often the host re-renders', () => {
+    const tile = renderUnderHostChurn(PROVIDER_CONFIG_SCHEMA);
+
+    // Load-bearing: proves the memo really ran its enrichment branch (two
+    // declared columns → two calls). Without it the assertion below would be
+    // vacuously green if `buildFieldMeta` were never reached at all.
+    const afterMount = counters.buildFieldMeta;
+    expect(afterMount).toBe(2);
+    // And that this is the discarding window — the table is not rendered.
+    expect(tile.queryByTestId('table')).toBeNull();
+    expect(tile.getByTestId('table-empty-state')).toBeTruthy();
+
+    tile.churn(3);
+
+    // Pre-fix this was 8: a fresh `[]` re-keyed `derivedColumns` on each of
+    // the three host renders, re-deriving both columns every time.
+    expect(counters.buildFieldMeta).toBe(afterMount);
+  });
+
+  it('still re-derives when rows actually arrive — the memo was stabilized, not frozen', () => {
+    let setData: ((d: any) => void) | null = null;
+    const Host = () => {
+      const [data, set] = React.useState<any>({ provider: 'object', object: 'account' });
+      setData = set;
+      return (
+        <ObjectDataTable
+          schema={{ type: 'object-table', data, columns: ['name', 'amount'] } as any}
+        />
+      );
+    };
+    const { getByTestId, queryByTestId } = render(
+      <I18nProvider config={I18N_CONFIG}><Host /></I18nProvider>,
+    );
+    expect(queryByTestId('table')).toBeNull();
+    const beforeRows = counters.buildFieldMeta;
+
+    act(() => { setData!([{ name: 'INV-1', amount: 100 }]); });
+
+    expect(counters.buildFieldMeta).toBeGreaterThan(beforeRows);
+    expect(getByTestId('table').textContent).toContain('INV-1');
+  });
+});

--- a/packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx
+++ b/packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx
@@ -1,0 +1,114 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * objectui#4629 (package sweep) — the identical expression in
+ * `ObjectPivotTable`.
+ *
+ * `const finalData = Array.isArray(rawData) ? rawData : []` evaluated a FRESH
+ * array literal on every render, and unlike `ObjectDataTable` this component
+ * has no unconditional empty-state early return: it hands the value straight
+ * to `PivotTable` as `finalSchema.data`, where it is the identity key of the
+ * cross-tabulation memo (`[data, rowField, columnField, valueField,
+ * aggregation]`). So the churn ESCAPED the component here and rebuilt the
+ * row/column sets, the bucket map and the totals on every render over nothing.
+ *
+ * This is the direct identity assertion: the same array object must reach
+ * `PivotTable` on every render. An `eslint-disable` suppression would not move
+ * it, and it is not satisfiable by "the pivot renders correctly" (which is
+ * green against the broken code too).
+ */
+import { describe, it, expect, afterEach, vi } from 'vitest';
+import { render, cleanup, act } from '@testing-library/react';
+import React from 'react';
+
+const captured = vi.hoisted(() => ({ data: [] as any[] }));
+
+vi.mock('../PivotTable', () => ({
+  PivotTable: ({ schema }: any) => {
+    captured.data.push(schema.data);
+    return <div data-testid="pivot">{(schema.data ?? []).map((r: any, i: number) => <span key={i}>{String(r.stage)}</span>)}</div>;
+  },
+}));
+
+import { ObjectPivotTable } from '../ObjectPivotTable';
+
+afterEach(() => {
+  cleanup();
+  captured.data.length = 0;
+});
+
+/**
+ * `data` is a PROVIDER CONFIG object rather than rows — the truthy non-array
+ * that selects the fallback arm. No `objectName`, so no fetch and no
+ * "no data source" early return: the component reaches `PivotTable`.
+ */
+const PROVIDER_CONFIG_SCHEMA = {
+  type: 'pivot-table',
+  rowField: 'stage',
+  columnField: 'owner',
+  valueField: 'amount',
+  data: { provider: 'object', object: 'opportunity' },
+} as any;
+
+function renderUnderHostChurn(schema: any) {
+  let bump: (() => void) | null = null;
+  const Host = () => {
+    const [, setTick] = React.useState(0);
+    bump = () => setTick((n) => n + 1);
+    return <ObjectPivotTable schema={schema} />;
+  };
+  const utils = render(<Host />);
+  return {
+    ...utils,
+    churn: (times: number) => {
+      for (let i = 0; i < times; i += 1) act(() => { bump!(); });
+    },
+  };
+}
+
+describe('ObjectPivotTable keeps "no rows yet" stable (#4629)', () => {
+  it('hands PivotTable one and the same empty array across host re-renders', () => {
+    const tile = renderUnderHostChurn(PROVIDER_CONFIG_SCHEMA);
+
+    // Load-bearing: the component really reached `PivotTable`, and with "no
+    // rows" — otherwise the identity assertion below would be vacuous.
+    expect(captured.data.length).toBe(1);
+    expect(captured.data[0]).toEqual([]);
+
+    tile.churn(3);
+
+    expect(captured.data.length).toBeGreaterThanOrEqual(4);
+    // Pre-fix: one distinct identity per render (4). The whole fix is that
+    // this is 1.
+    expect(new Set(captured.data).size).toBe(1);
+    // The module-scope empty is frozen, so a consumer that mutates the array
+    // it was handed cannot corrupt every other pivot on the page.
+    expect(Object.isFrozen(captured.data[0])).toBe(true);
+  });
+
+  it('still forwards rows when they arrive — the pass-through was not removed', () => {
+    let setData: ((d: any) => void) | null = null;
+    const Host = () => {
+      const [data, set] = React.useState<any>({ provider: 'object', object: 'opportunity' });
+      setData = set;
+      return (
+        <ObjectPivotTable
+          schema={{ ...PROVIDER_CONFIG_SCHEMA, data } as any}
+        />
+      );
+    };
+    const { getByTestId } = render(<Host />);
+    const rows = [{ stage: 'Won', owner: 'ann', amount: 10 }];
+
+    act(() => { setData!(rows); });
+
+    expect(captured.data[captured.data.length - 1]).toBe(rows);
+    expect(getByTestId('pivot').textContent).toContain('Won');
+  });
+});

--- a/packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx
+++ b/packages/plugin-dashboard/src/__tests__/ObjectPivotTable.stableEmptyRows.test.tsx
@@ -18,6 +18,11 @@
  * aggregation]`). So the churn ESCAPED the component here and rebuilt the
  * row/column sets, the bucket map and the totals on every render over nothing.
  *
+ * This closes the OBJECT-BOUND path only: the stable empty passes through
+ * `PivotTable`'s own `Array.isArray` arm unchanged. `PivotTable` still spells
+ * two per-render literals of its own for callers that reach it directly — out
+ * of this card's file surface, filed as objectui#5562.
+ *
  * This is the direct identity assertion: the same array object must reach
  * `PivotTable` on every render. An `eslint-disable` suppression would not move
  * it, and it is not satisfiable by "the pivot renders correctly" (which is

